launcher/instancemanager.go

```diff
@@ -0,0 +1,202 @@
+	removeServiceChannel   <-chan string
+}
+
+/***********************************************************************************************************************
+ * Private
+ **********************************************************************************************************************/
+
+func newInstanceManager(config *config.Config, storage Storage, storageStateProvider StorageStateProvider, removeServiceChannel <-chan string) (im *instanceManager, err error) {
+	im = &instanceManager{
+		config:                 config,
+		storage:                storage,
+		storageStateProvider:   storageStateProvider,
+		validateTTLStopChannel: make(chan struct{}),
+		removeServiceChannel:   removeServiceChannel,
+		uidPool:                uidgidpool.NewUserIDPool(),
+	}
+
+	im.fillUIDPool()
+
+	if err := im.clearInstancesWithDeletedService(); err != nil {
+		log.Errorf("Can't clear instances with deleted service: %v", err)
+
+		return nil, err
+	}
+
+	if err := im.removeOutdatedInstances(); err != nil {
+		log.Errorf("Can't remove outdated instances: %v", err)
+
+		return nil, err
+	}
+
+	go im.instanceRemover()
+
+	return im, nil
+}
+
+// acquireUID returns free UID from pool.
+func (im *instanceManager) acquireUID() (int, error) {
+	return im.uidPool.GetFreeID()
+}
+
+// close stops instance manager
+func (im *instanceManager) close() {
+	close(im.validateTTLStopChannel)
+}
+
+// fillUIDPool fills UID pool with UIDs from instances storage.
+func (im *instanceManager) fillUIDPool() {
+	instances, err := im.storage.GetInstances()
+	if err != nil {
+		log.Errorf("Can't fill UID pool: %v", err)
+	}
+
+	for _, instance := range instances {
+		if err = im.uidPool.AddID(instance.UID); err != nil {
+			log.Errorf("Can't add UID to pool: %v", err)
+		}
+	}
+}
+
+// releaseUID returns UID to pool
+func (im *instanceManager) releaseUID(uid int) error {
+	return im.uidPool.RemoveID(uid)
+}
+
+// removeOutdatedInstances removes cached instances with expired TTL.
+func (im *instanceManager) removeOutdatedInstances() error {
+	instances, err := im.storage.GetInstances()
+	if err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	for _, instance := range instances {
+		if !instance.Cached ||
+			time.Since(instance.Timestamp) < time.Hour*24*time.Duration(im.config.ServiceTTLDays) {
+
+			continue
+		}
+
+		if err := im.removeInstance(instance); err != nil {
+			// Log error and continue removing other outdated instances
+
+			log.Errorf("Can't remove instance: %v", err)
+		}
+	}
+
+	return nil
+}
+
+// removeInstance removes instance from storage and database. Returns instance's UID to pool.
+func (im *instanceManager) removeInstance(instanceInfo InstanceInfo) error {
+	err := im.storageStateProvider.RemoveServiceInstance(instanceInfo.InstanceIdent)
+	if err != nil && !errors.Is(err, ErrNotExist) {
+		return aoserrors.Wrap(err)
+	}
+
+	err = im.storage.RemoveInstance(instanceInfo.InstanceIdent)
+	if err != nil && !errors.Is(err, ErrNotExist) {
```

Comment:
```suggestion
	if err = im.storage.RemoveInstance(instanceInfo.InstanceIdent); err != nil && !errors.Is(err, ErrNotExist) {
```

if err was used before you can write it in one line